From 90349475a6fcbf874623b5de959326059fd31e16 Mon Sep 17 00:00:00 2001
From: Alexander Egorenkov <egorenar-dev@posteo.net>
Date: Sat, 17 Apr 2021 18:51:05 +0200
Subject: [PATCH 1/3] lib: scatterlist: Fix SGL length in sg_split() if
 !CONFIG_NEED_SG_DMA_LENGTH

If CONFIG_NEED_SG_DMA_LENGTH is NOT enabled then sg_dma_len() is an alias
for the length field in a SGL. In that case sg_split() wrongly resets
the length of split SGLs to zero after it was set correctly before.

Signed-off-by: Alexander Egorenkov <egorenar-dev@posteo.net>
---
 lib/sg_split.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/lib/sg_split.c
+++ b/lib/sg_split.c
@@ -60,15 +60,16 @@ static int sg_calculate_split(struct sca
 			curr->length_last_sg = len;
 			size -= len;
 		}
+
+		if (!nb_splits)
+			break;
+
 		skip = 0;
 
 		if (!size && --nb_splits > 0) {
 			curr++;
 			size = *(++sizes);
 		}
-
-		if (!nb_splits)
-			break;
 	}
 
 	return (size || !splitters[0].in_sg0) ? -EINVAL : 0;
@@ -88,11 +89,10 @@ static void sg_split_phys(struct sg_spli
 			if (!j) {
 				out_sg->offset += split->skip_sg0;
 				out_sg->length -= split->skip_sg0;
-			} else {
-				out_sg->offset = 0;
 			}
 			sg_dma_address(out_sg) = 0;
-			sg_dma_len(out_sg) = 0;
+			if (IS_ENABLED(CONFIG_NEED_SG_DMA_LENGTH))
+				sg_dma_len(out_sg) = 0;
 			in_sg = sg_next(in_sg);
 		}
 		out_sg[-1].length = split->length_last_sg;
